### PR TITLE
feat: migrate OnboardingController to BaseController v2

### LINKED ISSUE #25927: Migrate OnboardingController to BaseController v2
### What is this about?

Following the [Wallet Framework team's OKRs for Q3 2024](https://docs.google.com/document/d/1JLEzfUxHlT8lw8ntgMWG0vQb5BAATcrYZDj0wRB2ogI/edit#heading=h.kzzai3cfecro), we want to bring `OnboardingController` up to date with our latest controller patterns.

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

(TODO)

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

- https://github.com/MetaMask/core/issues/1509

### PR DESCRIPTION
**Pull request by cryptodev-2s**
## **Description**

This PR updates `OnboardingController` to inherit from BaseController V2.

Details about other changes are to come...

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/26880?quickstart=1)

## **Related issues**

Fixes: [#25927](https://github.com/MetaMask/metamask-extension/issues/25927)

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Review by mikesposito — COMMENTED** [review]
This looks good overall! I just added a couple of small comments

File: app/scripts/controllers/onboarding.ts [hunk 1]
```diff
@@ -12,41 +20,125 @@ export type OnboardingControllerState = {
...
-      ...initState,
-      ...defaultTransientState,
+    super({
+      messenger,
+      metadata: controllerMetadata,
+      name: controllerName,
+      state: {
+        ...getDefaultOnboardingControllerState(),
+        ...state,
+        ...defaultTransientState,
```
[the remark this answers is not in the document]
Comment (mikesposito) [reply app/scripts/controllers/onboarding.ts:140]: Oh, I see, this is not supposed to be passed by the consumer. Though, does this mean that we shouldn't persist? (`persist: false` in metadata)

[the remark this answers is not in the document]
Comment (cryptodev-2s) [reply app/scripts/controllers/onboarding.ts:140]: I am not sure but I mostly think it shouldn't so I just did update it

File: app/scripts/lib/account-tracker.test.js [hunk 2]
```diff
@@ -94,6 +95,17 @@ function buildAccountTracker({
...
+  const onboardingController = new OnboardingController({
+    messenger: controllerMessenger.getRestricted({
+      name: 'OnboardingController',
+      allowedActions: [],
+      allowedEvents: [],
+    }),
+    state: {
+      completedOnboarding,
+    },
+  });
```
[the remark this answers is not in the document]
Comment (cryptodev-2s) [reply app/scripts/lib/account-tracker.test.js]: Yes as we only need the onboarding state I am just mocking the state

File: app/scripts/controllers/onboarding.ts [hunk 3]
```diff
@@ -12,41 +20,125 @@ export type OnboardingControllerState = {
   onboardingTabs?: Record<string, string>;
 };
 
+/**
+ * Function to get default state of the {@link OnboardingController}.
+ */
+export const getDefaultOnboardingControllerState = () => ({
+  seedPhraseBackedUp: null,
+  firstTimeFlowType: null,
```
[the remark this answers is not in the document]
Comment (cryptodev-2s) [reply app/scripts/controllers/onboarding.ts:28]: Yes I believe this might require a migration indeed can also be done later if it make sense

**Review by mikesposito — APPROVED** [review]
LGTM!

### **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
  - ✅ Tested onboarding from fresh install
  - ✅ Tested with an already onboarded account 
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

**Review by itsyoboieltr — APPROVED** [review]
LGTM!

**Pull request merged**